Require all packages to solve / compile and include all valid compilers in their metadata

thomashoneyman opened the pull request:
Fixes #577. Fixes #255.

The core problem solved here is identifying what compilers are compatible with a specific package version, such as `aff@7.0.0`. We need this to support an oft-requested feature for Pursuit: filtering search results by a compiler version (or range). It's also useful for other things; for one, it allows us to add the compiler version as a constraint to the solver to produce a build plan that works with the specific compiler given.

Metadata files now include a `compilers` key in published metadata that lists either a bare version (the version used to publish the package) or an array of versions (the full set of compilers known to work with the package). The reason for two representations is that computing the full set of compilers can take a long time; this approach lets us upload the package right away and compute the rest of the valid compilers in a fixup pass. A bare version means the full set has not been computed yet.

All packages must now be solvable. We can't compile a package version if we can't fetch its dependencies, so this becomes a requirement for all packages.

There are only 2 scenarios in which we need to compute the available compilers for a package version:

1. A new package version is published
2. A new compiler version is published

This PR is focused on the first case, and we should do a followup for the second case. (The second case is straightforward, and @colinwahl's compiler versions script already essentially implements it. It's been omitted from this PR for brevity).

A new package version can be published via the legacy importer or via a user submitting an API call, but the result is the same: eventually the `publish` pipeline is run. For that reason I've decided to compute the compiler versions for a package version as part of the publish pipeline where we're already determining resolutions and building with a specific compiler. That centralizes the logic to a single place.

Therefore this PR centers on two things: trying compilers to find all that work for a package version at the end of publishing, and updating the legacy importer to determine a valid compiler version and resolutions before calling `publish`.

I've added some tests and I've run the legacy importer locally; it's about 500 packages in so far and every failure appears to be correct. More comments in the PR.

thomashoneyman commented on app/src/App/API.purs:
```diff
@@ -504,20 +515,30 @@ publish source payload = do
       Right versions -> pure versions
 
     case Map.lookup manifest.version published of
+      Nothing | payload.compiler < unsafeFromRight (Version.parse "0.14.7") -> do
```
`purs publish` will fail for packages prior to 0.14.7 because that's when we added support for the purs.json file format. Before that the compiler looks for specific Bowerfile fields that aren't present in the purs.json file. Since all these packages ought to already be published to Pursuit I think this is fine. We can't do anything about it anyway until #525.

thomashoneyman commented on app/src/App/API.purs:
````diff
@@ -734,50 +743,87 @@ publishRegistry { source, payload, metadata: Metadata metadata, manifest: Manife
+    , "\n```"
+    ]
 
-  case compilationResult of
-    Left error
-      -- We allow legacy packages to fail compilation because we do not
-      -- necessarily know what compiler to use with them.
-      | source == LegacyPackage -> do
-          Log.debug error
-          Log.warn "Failed to compile, but continuing because this package is a legacy package."
-      | otherwise ->
-          Except.throw error
-    Right _ ->
-      pure unit
````
This code is no longer needed because packages must compile.

thomashoneyman commented on app/src/App/API.purs:
```diff
@@ -734,50 +743,87 @@ publishRegistry { source, payload, metadata: Metadata metadata, manifest: Manife
+    , ". Computing the list of compilers usable with your package version..."
+    ]
+
+  { failed: invalidCompilers, succeeded: validCompilers } <- findAllCompilers
+    { source: packageDirectory
+    , installed: installedResolutions
+    , compilers: Array.fromFoldable $ NonEmptySet.filter (notEq payload.compiler) compatible
+    }
```
This is where we determine all valid compilers — at the end of publishing, after the metadata upload has already happened. I did it this way in case we want to fork this off as a separate job so publishing is faster. 

However, if we just want this to be part of publishing then we could do this before uploading metadata. That would let us avoid the two states in metadata (bare version vs. array of versions). Instead we'd always just go straight to the array of versions.

thomashoneyman commented on app/src/App/API.purs:
```diff
@@ -851,55 +897,119 @@ validateResolutions manifest resolutions = do
+
+-- | Given a set of package versions, determine the set of compilers that can be
+-- | used for all packages.
+compatibleCompilers :: Map PackageName Metadata -> Map PackageName Version -> Maybe (NonEmptySet Version)
+compatibleCompilers allMetadata resolutions = do
```
Basically, the idea here is that we've machine-checked every dependency and know what compilers it can possibly work with. If a dependency doesn't work with `0.15.0` then your package won't either; the dependency will fail to compile. So we don't have to try _every_ compiler version, just those that dependencies work with.

thomashoneyman replied:
I've taken this out because it's not quite correct. Instead, we want to re-solve the package every time we choose a different compiler in case the solution is different. We can only do that because of the changes in this PR — having the compilers in metadata.

thomashoneyman commented on app/src/App/Effect/Cache.purs:
```diff
@@ -168,7 +168,6 @@ handleMemoryFs env = case _ of
           case inFs of
             Nothing -> pure $ reply Nothing
             Just entry -> do
-              Log.debug $ "Fell back to on-disk entry for " <> memory
```
These are just _so_ noisy. Maybe we can introduce a `Log.superDebug`.

thomashoneyman replied:
Yea, I think they're not really useful now that we're confident the cache works correctly. I had them in there from when I first developed it and would either sometimes see things I thought should be cached _not_ get cached, or I wanted to make sure something I removed from the cache really was.

thomashoneyman commented on scripts/src/LegacyImporter.purs:
```diff
@@ -211,102 +219,119 @@ runLegacyImport mode logs = do
-            }
-
-  case notPublished of
+    publishLegacyPackage :: Manifest -> Run _ Unit
+    publishLegacyPackage (Manifest manifest) = do
```
This is where we solve, compile, and then publish each package in turn. Publish failures are saved to cache and, at the end of the process, written to a `publish-failures.json` file that records every version that failed and its reason so we can hand-review it. I've run this on a few hundred packages and it's looking correct.

thomashoneyman reviewed (commented):
I've included a few review comments that describe how various parts of the code work. But I'm also happy to jump on a call in the PureScript chat to walk through this and answer any questions.

thomashoneyman commented:
An example from the publish-failures.json file:

```json
{
  "string-parsers": {
    "3.0.1": {
      "reason": "No versions found in the registry for lists in range\n  >=4.0.0 (declared dependency)\n  <5.0.0 (declared dependency)",
      "tag": "SolveFailed"
    },
    "3.1.0": {
      "reason": "No versions found in the registry for lists in range\n  >=4.0.0 (declared dependency)\n  <5.0.0 (declared dependency)",
      "tag": "SolveFailed"
    }
  },
  "strings": {
    "3.5.0": {
      "reason": "No valid compilers found for publishing.",
      "tag": "NoCompilersFound",
      "value": {
        "0.13.0,0.13.2,0.13.3,0.13.4,0.13.5,0.13.6,0.13.8": [
          {
            "errorCode": "ModuleNotFound",
            "errorLink": "https://github.com/purescript/documentation/blob/master/errors/ModuleNotFound.md",
            "filename": "/tmp/nix-shell.s8MfBC/nix-shell.IqkUN7/tmp-1515056-wZ6mar1q4MJl/purescript-strings/src/Data/String/CodePoints.purs",
            "message": "  Module Data.Int was not found.\n  Make sure the source file exists, and that it has been provided as an input to the compiler.\n",
            "moduleName": "Data.String.CodePoints",
            "position": {
              "endColumn": 42,
              "endLine": 34,
              "startColumn": 1,
              "startLine": 34
            }
          }
        ]
      }
    }
  },
}
```

thomashoneyman commented:
I've uncovered two rare but significant issues affecting the legacy importer (unrelated to this PR).

**Topological sorting**
First: in some cases our typical approach of topologically sorting manifests by their dependencies (ignoring their explicit bounds) will fail to read a valid index from disk. This happens when a package like `functors` at one time depends on another (like `contravariant`), but at other times the dependency is flipped. Since we don't take specific version ranges into consideration, we're at the mercy of other conflicting sorting orders in the map as to whether these end up going in the right order or not. Here's the output of a failed run reading an index from disk:

```
success: 'Inserted distributive@5.0.0'
success: 'Inserted exists@5.1.0'
success: 'Inserted exists@5.0.0'
success: 'Inserted profunctor@5.0.0'
fail: 'Failed to insert functors@4.1.1: \n  - contravariant>=5.0.0 <6.0.0'
success: 'Inserted functors@3.1.1'
success: 'Inserted functors@3.1.0'
success: 'Inserted functors@3.0.1'
success: 'Inserted functors@3.0.0'
success: 'Inserted foldable-traversable@5.0.0'
success: 'Inserted either@4.1.0'
success: 'Inserted foldable-traversable@5.0.1'
success: 'Inserted either@4.1.1'
success: 'Inserted contravariant@4.0.1'
success: 'Inserted const@4.0.0'
success: 'Inserted contravariant@5.0.0'
```

This run fails to produce a valid index because functors@4.1.1 is unsatisfied in its dependency on contravariant, but of course we see contravariant get inserted a little later on.

The solution is to always consider version bounds when reading an index from disk where we expect bounds to be at least reasonably correct. I've implemented and tested that and situations like this no longer happen.

The reason we ignored ranges at first is because we had far fewer checks around correct bounds and because in the package sets we _want_ to explicitly ignore ranges when working with an index (when doing, for example, the 'self-contained' check). I've preserved that behavior — you can always opt-in to either considering or ignoring ranges when working with an index.

**Incorrect dependencies detected in legacy manifests derived from package sets**

Second, in some cases a package like strings@3.5.0 will have its dependency list pruned to only those listed in its package sets entry, and those turn out to be overly-restrictive. In this case, specifically, the dependency on `purescript-integers` listed in its Bowerfile is removed because there is no such dependency in its package sets list; in the package sets that dependency ended up being picked up transitively, but when we go to solve and compile the package the solution picks a different transitive dependency and the package fails.

This shouldn't happen because strings@3.5.0 has a Bowerfile that explicitly lists a dependency on integers, which we trimmed out by deferring to the package sets entry. We did this assuming package sets entries are correct and because we didn't want overly-constrained dependency lists.

The second concern is no longer valid because with #667 we will remove unused dependencies. The first concern is no longer reasonable because we have at least one example of package sets dependency lists being incorrect.

The solution is simple: instead of preferring package sets entries over other manifests, just union them all and defer to the 'unused dependencies' pruning in the publishing pipeline to trim out ones that aren't actually needed.

thomashoneyman commented on scripts/src/LegacyImporter.purs:
```diff
@@ -211,102 +219,128 @@ runLegacyImport mode logs = do
-        let metadata = Metadata { location, owners: Nothing, published: Map.empty, unpublished: Map.empty }
-        Registry.writeMetadata package metadata
-      Just _ -> pure unit
+  let metadataPackage = unsafeFromRight (PackageName.parse "metadata")
+  Registry.readMetadata metadataPackage >>= case _ of
+    Nothing -> do
+      Log.info "Writing empty metadata file for the 'metadata' package"
+      let location = GitHub { owner: "purescript", repo: "purescript-metadata", subdir: Nothing }
+      let entry = Metadata { location, owners: Nothing, published: Map.empty, unpublished: Map.empty }
+      Registry.writeMetadata metadataPackage entry
+    Just _ -> pure unit
```
We agreed to not reserve package names pre-0.13.0, so this reserves only metadata for the "metadata" package used by legacy package sets.

thomashoneyman commented on scripts/src/LegacyImporter.purs:
```diff
@@ -211,102 +219,128 @@ runLegacyImport mode logs = do
-        , Array.foldMap (append "\n  - " <<< printPackage) manifests
         ]
 
+      void $ for (Array.take 500 manifests) publishLegacyPackage
```
(this is for testing's sake)

```suggestion
      void $ for manifests publishLegacyPackage
```

thomashoneyman commented:
Can confirm that the fix is working with regards to generating manifests with full dependency lists to be pruned later — this is `strings@3.5.0`, for example:

```json
{
  "name": "strings",
  "version": "3.5.0",
  "license": "MIT",
  "description": "String and char utility functions, regular expressions.",
  "location": {
    "githubOwner": "purescript",
    "githubRepo": "purescript-strings"
  },
  "dependencies": {
    "arrays": ">=4.0.1 <5.0.0",
    "either": ">=3.0.0 <4.0.0",
    "gen": ">=1.1.0 <2.0.0",
    "integers": ">=3.2.0 <4.0.0",
    "maybe": ">=3.0.0 <4.0.0",
    "partial": ">=1.2.0 <2.0.0",
    "unfoldable": ">=3.0.0 <4.0.0"
  }
}
```

thomashoneyman commented:
We also need to support spago.yaml files in the legacy importer, as some packages now use that format. Otherwise they will be excluded with a `NoManifests` error.

thomashoneyman commented:
Here's another fun one: some packages, like `transformers@3.6.0`, list dependencies which are entirely unused, in this case `arrays`. We then prune this dependency. However, then the package fails because it brings in _other_ dependencies via that transitive dependency, such as `either`.

That means we can't get away with simply removing unused dependencies because we may also remove direct imports that were pulled in transitively by the unused dependency. Either we give up on removing unused dependencies, or we both remove unused dependencies and insert direct imports that weren't mentioned.

For dependencies we insert into a manifest we have their exact version via the solver; we could potentially do a `bumpHighest` on them to produce a range the same way we do when working with spago.dhall files.

I think that's preferable to giving up on the unused dependencies check but I'm curious if you disagree @f-f or @colinwahl.

thomashoneyman commented:
As of the latest commit: we no longer simply remove unused dependencies. Instead, we loop. We remove unused dependencies, then bring in any transitive dependencies they would have brought in, and then check the new dependency list for unused dependencies and so on.

The result is that we remove unused dependencies while preserving any transitive dependencies they brought it which are used in the source code. Note that we don't go through and add _all_ packages your code directly imports, we just do this for dependencies that are being removed.

thomashoneyman commented:
@f-f this is ready for a look. In my test run I'm thousands of package versions in and everything looks good. Packages are building and we're recording the full set of compiler versions for them. Solver and compilation failures look correct. We're pruning unused dependencies from packages that list them (replacing the removed deps with their transitives and pruning again).

I took a stab at automatically adding directly-imported packages to the manifest of packages that don't list them, but it's funky. We can't be sure we're accurately reflecting the ranges they intended. I can cover most cases but I don't at all feel confident I can catch them all. It's not technically incorrect for a user to omit packages they directly import from their manifest if they get brought in transitively, and unlike the "unused dependency" check we put in place they're not going to pull in random test and dev dependencies.

Anyways — at this point the legacy importer and publish pipeline are solving and compiling packages as they publish and I think this is good to go. I'll share the full publish failure stats when I have a completed run of the importer.

thomashoneyman commented on scripts/src/LegacyImporter.purs:
```diff
@@ -211,102 +228,184 @@ runLegacyImport mode logs = do
 
-  Log.info "Writing empty metadata files for legacy packages that can't be registered..."
-  void $ forWithIndex importedIndex.reservedPackages \package location -> do
+  let metadataPackage = unsafeFromRight (PackageName.parse "metadata")
+  let pursPackage = unsafeFromRight (PackageName.parse "purs")
```
We only reserve the "metadata" and "purs" packages, as discussed, releasing names that have no published versions.

thomashoneyman replied:
Might need to reconsider this given the number of packages being dropped.

f-f replied:
We need these to be considered separately right? I.e. they can't be in the section above with the `filterPackages_0_13` - we have versions for `metadata` but we want none, and `purs` hasn't been published at all.
We should probably reserve `purescript` too?

thomashoneyman commented on lib/src/Solver.purs:
```diff
@@ -39,6 +48,50 @@ import Safe.Coerce (coerce)
+      -- then we don't add it to the dependencies to avoid over-
+      -- constraining the solver.
+      compilers <- Either.hush eitherCompilers
+      -- Otherwise, we construct a maximal range for the compilers the
```
It's a potential issue in that you could have packages A, B, and C such that:

* A supports `0.15.0`, `0.15.3`, `0.15.4` or range `>=0.15.0 <0.15.5`
* B supports `0.15.1`, `0.15.3`, `0.15.4` or range `>=0.15.1 <0.15.5`
* C supports `0.15.0`, `0.15.2`, `0.15.3` or range `>=0.15.0 <0.15.4`

and the resulting restricted range would be `>=0.15.1 <0.15.4` even though two of the three compilers that satisfy those ranges (`0.15.1` and `0.15.2`) are definitely broken (A and C don't accept `0.15.1`, A and B don't accept `0.15.2`). Therefore it's technically possible to solve for something that definitely doesn't work given the input compilers.

That's the worst case problem. In the context of this PR that's no issue because it just means we waste a little time trying compilers that won't work. It won't affect correctness.

I don't think it's too big of a problem in the wild, either. I suspect ranges will regularly include versions that don't actually work (some patch version is broken within the range) and we can't guarantee that they're correct. Just that our plan is correct given the ranges you gave us. Our compiler ranges may sometimes be incorrect if a package works across a range of compilers where some don't work in the middle, and I think that's acceptable.

thomashoneyman replied:
(cc @f-f, @monoidmusician)

colinwahl commented on app/src/App/API.purs:
```diff
@@ -743,73 +696,100 @@ publishRegistry { source, payload, metadata: Metadata metadata, manifest: Manife
+  Comment.comment "Mirrored registry operation to the legacy registry!"
+
+  Log.debug "Uploading package documentation to Pursuit"
+  if payload.compiler >= unsafeFromRight (Version.parse "0.14.7") then
```
There are a few instances of `unsafeFromRight (Version.parse "0.14.7")` within this file, all denoting the minimum compiler version required to upload docs to pursuit - should we pull that out into a constant?

thomashoneyman replied:
I've moved the minimums out to the Purs module as constants. thanks!

thomashoneyman commented:
Statistics:

* 1225 out of 1,440 packages had at least one version fail
   * 215 had all versions fail and will be removed
* 8394 out of 10,682 versions failed
   * 7322 solving failed (dependency versions missing, ie. old packages)
   * 537 no compilers usable (all tried, but compilation failed)
   * 473 solving failed (compilers are not compatible)
   * 62 publishing failed (something went wrong getting the package)

[publish-failures.json](https://github.com/purescript/registry-dev/files/13529311/publish-failures.json)

The 62 failures look like they are either intermittent `git clone` errors that persist through multiple retries, or happened because the registry was updated while the import was running. I'm going to clear out those 62 failures and retry when I have internet access restored at home. I will also add `unsafeCrash` calls to the importer when we do the full reimport so these are caught and fixed immediately.

In sum: this looks good & correct to me, and I think that — pending review — we can move forward with reimporting the registry either this weekend or next weekend. During that period of time it will be necessary to disable package publishing so no interruptions happen.

colinwahl commented on app/src/App/API.purs:
```diff
@@ -743,73 +696,100 @@ publishRegistry { source, payload, metadata: Metadata metadata, manifest: Manife
+      ]
+
+  allCompilers <- PursVersions.pursVersions
+  { failed: invalidCompilers, succeeded: validCompilers } <- case NonEmptyArray.fromFoldable $ NonEmptyArray.filter (notEq payload.compiler) allCompilers of
```
Could this be `case NonEmptyArray.fromFoldable $ NonEmptyArray.delete payload.compiler allCompilers of`?

thomashoneyman replied:
Fixed!

colinwahl commented on app/src/App/API.purs:
```diff
@@ -743,73 +696,100 @@ publishRegistry { source, payload, metadata: Metadata metadata, manifest: Manife
+    Log.debug $ "Some compilers failed: " <> String.joinWith ", " (map Version.print (Set.toUnfoldable (Map.keys invalidCompilers)))
+
+  let
+    allVerified = case NonEmptySet.fromFoldable validCompilers of
```
`validCompilers` is already constructed to always contain `payload.compilers` above - are these 3 lines necessary?

thomashoneyman replied:
Good point! I've addressed that.

colinwahl commented on app/src/App/API.purs:
```diff
@@ -860,55 +840,113 @@ validateResolutions manifest resolutions = do
+            FS.Extra.ensureDirectory installed
+            installBuildPlan resolutions installed
+            result <- Run.liftAff $ Purs.callCompiler
+              { command: Purs.Compile { globs: [ Path.concat [ source, "src/**/*.purs" ], Path.concat [ installed, "*/src/**/*.purs" ] ] }
```
`*/src/**/*.purs` appears throughout this file as a glob to include the source files of all dependencies - should that be a constant?

thomashoneyman replied:
I tried pulling it out as a constant but I find it personally easier to understand when I can see the exact globs being used. If you feel strongly about this then I can make the change, though.

colinwahl commented on app/src/App/API.purs:
```diff
@@ -954,38 +992,37 @@ parseInstalledModulePath { prefix, path } = do
--- | packages where the `purescript-` prefix is still present.
+-- | packages where the `purescript-` prefix is still present. Cannot be used
+-- | on packages prior to 0.14.7.
 publishToPursuit
```
I know elsewhere when we call this function we've already checked the compiler version, but would it be easier to push those checks into this function itself to avoid duplication & ensure we actually don't call it with packages prior to compiler version 0.14.7?

thomashoneyman replied:
I've kept that check outside because it lets us curate the error messages (it's not really an _error_ if you publish with a pre-0.14.7 compiler, we'll just tell you we're skipping publishing, whereas actually attempting it would produce an error.) I've put an error check in there just to be sure, though.

colinwahl commented on app/src/App/API.purs:
```diff
@@ -1162,3 +1184,175 @@ getPacchettiBotti = do
+  -- Technically that's 0.13.8, but that version had a bug wrt transitive
+  -- dependencies, so we start from 0.14.0.
+  let minGraphCompiler = unsafeFromRight (Version.parse "0.14.0")
+  let compiler' = if compiler >= minGraphCompiler then compiler else minGraphCompiler
```
Just trying to make sure I have everything straight in my head...

Here, we can replace the payload compiler version with `minGraphCompiler` safely because `purs graph` doesn't actually compile anything, it just reads module headers and produces a module graph?

I think this has me confused because below we catch `CompilationError` - I want to make sure we don't introduce spurious compilation errors by changing the requested payload compiler with a future compiler that is not expected to work.

thomashoneyman replied:
We're within a function that adjusts the dependencies in a manifest based on the output of purs graph. If their compiler version is too low to support purs graph then we use the minimum compiler that does support it instead, but if that fails, then we just bail out and don't adjust the dependencies. (This doesn't bail out of publishing — just out of adjusting their dependencies.) Sounds like, from your comment below, you saw this code path.

colinwahl commented on app/src/App/API.purs:
```diff
@@ -1162,3 +1184,175 @@ getPacchettiBotti = do
+        -- package uses syntax before the oldest usable purs graph compiler (ie.
+        -- 0.14.0). In this case we can't determine unused dependencies and should
+        -- leave the manifest untouched.
+        pure $ Tuple (Manifest manifest) verified
```
Oh, nevermind, I think my previous comment is addressed here.

thomashoneyman commented:
Output of the test run:

```
--------------------
PUBLISH FAILURES
--------------------

PACKAGES: 1222 out of 1440 had at least 1 version fail (755 packages have zero usable versions).

VERSIONS: 8388 out of 10686 failed.
  - Publishing failed: 25
  - Solving failed (compiler): 473
  - No compilers usable for publishing: 538
  - Solving failed (dependencies): 7352
```

Here is the full list of failures — the vast majority are because of missing dependencies, ie. the package refers to some package version at a range that doesn't exist, and that's because many package versions are very old.

[publish-failures.json](https://github.com/purescript/registry-dev/files/13531667/publish-failures.json)

In addition, here is the full list of packages that had no usable versions at all and will be removed and their names released for others to register:

[removed-packages.txt](https://github.com/purescript/registry-dev/files/13531689/removed-packages.txt)

thomashoneyman commented:
It looks like some of these failures are because I had cached failures from my many, many test runs of the script. I have fully cleared the cache and am re-running from scratch and should be able to report the results tomorrow. Already I can see a good number of additional packages succeeding.

thomashoneyman commented:
Results of the run without any cached errors — significant increase in the number of package versions kept, but still many packages without usable versions.

```
--------------------
PUBLISH FAILURES
--------------------

PACKAGES: 999 out of 1440 had at least 1 version fail (515 packages have zero usable versions).
VERSIONS: 6475 out of 10691 failed.

  - Publishing failed: 33
  - Solving failed (compiler): 100
  - No compilers usable for publishing: 628
  - Solving failed (dependencies): 5714
```

These are all of the failures:

[publish-failures.json](https://github.com/purescript/registry-dev/files/13540773/publish-failures.json)

These are the packages without any usable versions:

[removed-packages.txt](https://github.com/purescript/registry-dev/files/13540774/removed-packages.txt)

thomashoneyman commented:
Encountered another issue: sometimes we remove an unused dependency (as per #667) from a package (for example, we remove `record` as a dependency from `variant@8.0.0`) but this causes a failure in another package because _that_ one was bringing in the dependency transitively (for example, `codec-argonaut@9.2.0` uses `record` and was bringing it in transitively via `variant`).

In short, we can't just remove unused dependencies and insert any transitive dependencies that were being brought in by the unused package as described in https://github.com/purescript/registry-dev/pull/669/#issuecomment-1817062736. Some other package downstream may have been relying on the transitive dependencies.

We either have to:

1. Give up on pruning unused dependencies and have a policy that we accept transitive dependencies (and we should be clear that removing a dependency = a breaking change, even if the dependency was unused, because other packages may be relying on it)
2. Fix manifests not just by removing unused dependencies and inserting anything they brought in transitively, but by inserting _all_ directly-imported packages as dependencies in the manifest.

For (2) I have not yet come up with a robust method. However, @natefaubion suggested in the [PureScript chat](https://purescript.org/chat) that we could:

- Take the transitive dependencies of the untrimmed dependencies
- Take the transitive dependencies of the trimmed dependencies
- See what the diff is and add the discovered transitive dependency

thomashoneyman commented:
In this latest iteration I add support for inserting transitive dependencies as well as removing unused ones. This relies on keeping the untrimmed legacy manifests in a `TransitivizedRegistry` so we can use it for solving when discovering missing package ranges. Here's how we solve:

1. First we solve the untrimmed manifest using `solveSteps` on the legacy registry. This produces ranges for the manifest dependencies (and their dependencies) up to the point where the solver would have to make a commitment, and then it halts.
2. We look up the missing packages in the `solveSteps` result and take their ranges.
3. Then, we solve the untrimmed manifest again with `solveFull`. This produces exact versions for every dependency and transitive dependency in the manifest.
4. We iterate through every resolution's untrimmed manifest; if we see any of the remaining missing packages in its dependencies then we update the missing package's range.

The result is that most ranges are produced by the solver via `solveSteps`, which is more likely to be correct since it considers all the ranges that could be possible in a solution. However, some ranges will still be missing so we take the remaining ranges from the legacy resolutions' manifests.

The results seem good so far. For example, `enums@3.1.0` uses both sources of ranges:

First: we find some transitive dependencies to fix:

```
[2023-12-07T12:32:29.552Z INFO] Found fixable dependency errors: Missing dependencies (control, gen, maybe, newtype, nonempty, partial, prelude, tuples)
```

We produce legacy resolutions for the package:

```
[2023-12-07T12:32:29.674Z DEBUG] Got legacy resolutions:
{
  "arrays": "4.4.0",
  "bifunctors": "3.0.0",
  "control": "3.3.1",
  "distributive": "3.0.0",
  "eff": "3.2.3",
  "either": "3.2.0",
  "foldable-traversable": "3.7.1",
  "gen": "1.3.1",
  "globals": "3.0.0",
  "identity": "3.1.0",
  "integers": "3.2.0",
  "invariant": "3.0.0",
  "math": "2.1.1",
  "maybe": "3.1.0",
  "monoid": "3.3.1",
  "newtype": "2.0.0",
  "nonempty": "4.3.0",
  "partial": "1.2.1",
  "prelude": "3.3.0",
  "st": "3.0.0",
  "strings": "3.5.0",
  "tailrec": "3.3.0",
  "tuples": "4.1.0",
  "type-equality": "2.1.0",
  "unfoldable": "3.2.0",
  "unsafe-coerce": "3.0.0"
}
```

Then we take a look at the transitive solution (ie. using the transitivized registry)

```
[2023-12-07T12:32:29.772Z DEBUG] Got transitive solution:
{
  "bifunctors": ">=3.0.0 <4.0.0",
  "control": ">=3.0.0 <4.0.0",
  "either": ">=3.0.0 <4.0.0",
  "foldable-traversable": ">=3.0.0 <4.0.0",
  "invariant": ">=3.0.0 <4.0.0",
  "maybe": ">=3.0.0 <4.0.0",
  "monoid": ">=3.0.0 <4.0.0",
  "newtype": ">=2.0.0 <3.0.0",
  "partial": ">=1.2.0 <2.0.0",
  "prelude": ">=3.0.0 <4.0.0",
  "strings": ">=3.0.0 <4.0.0",
  "tuples": ">=4.0.0 <5.0.0",
  "unfoldable": ">=3.0.0 <4.0.0"
}
```

We're able to get most package ranges from here, but some are missing in this solution (`gen` and `nonempty`). So we continue, this time taking the ranges by looking at every package in the legacy resolutions to see if gen and nonempty are listed in their ranges, and produce the following result:

```
[2023-12-07T12:32:29.804Z INFO] [NOTIFY] Your package is using a legacy manifest format, so we have adjusted your dependencies to remove unused ones and add direct-imported ones. Your dependency list was:
{
  "either": ">=3.0.0 <4.0.0",
  "strings": ">=3.0.0 <4.0.0",
  "unfoldable": ">=3.0.0 <4.0.0"
}

We have added the following packages: control, gen, maybe, newtype, nonempty, partial, prelude, tuples

Your new dependency list is:
{
  "control": ">=3.0.0 <4.0.0",
  "either": ">=3.0.0 <4.0.0",
  "gen": ">=1.1.0 <2.0.0",
  "maybe": ">=3.0.0 <4.0.0",
  "newtype": ">=2.0.0 <3.0.0",
  "nonempty": ">=4.2.0 <5.0.0",
  "partial": ">=1.2.0 <2.0.0",
  "prelude": ">=3.0.0 <4.0.0",
  "strings": ">=3.0.0 <4.0.0",
  "tuples": ">=4.0.0 <5.0.0",
  "unfoldable": ">=3.0.0 <4.0.0"
}
```

So we get "gen": ">=1.1.0 <2.0.0" and "nonempty": ">=4.2.0 <5.0.0", which is probably a little restrictive (maybe if we'd had other resolutions we would have had gen: >=1.0.0 <2.0.0) but I think it's pretty harmless.

thomashoneyman commented on lib/src/Operation/Validation.purs:
```diff
@@ -71,11 +72,63 @@ isNotUnpublished :: Manifest -> Metadata -> Maybe UnpublishedMetadata
+  printPackages :: NonEmptySet PackageName -> String
+  printPackages = String.joinWith ", " <<< map PackageName.print <<< NonEmptySet.toUnfoldable
+
+-- | Verifies that the manifest lists dependencies imported in the source code,
+-- | no more (ie. unused) and no less (ie. transitive). The graph passed to this
+-- | function should be the output of 'purs graph' executed on the 'output'
+-- | directory of the package compiled with its dependencies.
+noTransitiveOrMissingDeps :: Manifest -> PursGraph -> (FilePath -> Either String PackageName) -> Either (Either (NonEmptyArray AssociatedError) ValidateDepsError) Unit
+noTransitiveOrMissingDeps (Manifest manifest) graph parser = do
```
This is now a check in the validation portion of the library (cc: @f-f) if package managers want to use it.

thomashoneyman commented:
Results of the run now that we fix dependencies in general:

```
--------------------
PUBLISH FAILURES
--------------------

999 out of 1443 packages had at least 1 version fail (516 packages had all versions fail).
6471 out of 10695 versions failed.

  - Publishing failed: 22
  - Solving failed (compiler): 104
  - No compilers usable for publishing: 628
  - Solving failed (dependencies): 5717
```

* Full list of publish failures with reasons: [publish-failures.json](https://github.com/purescript/registry-dev/files/13621320/publish-failures.json)
* Packages that had no successful versions: [removed-packages.txt](https://github.com/purescript/registry-dev/files/13621323/removed-packages.txt)
* Metadata produced by this run, with compilers associated: [metadata.zip](https://github.com/purescript/registry-dev/files/13621325/metadata.zip)
* Manifest index produced by this run, with dependencies all fixed: [registry-index.zip](https://github.com/purescript/registry-dev/files/13621333/registry-index.zip)

thomashoneyman commented:
Some packages (notably `codec-argonaut` and `codec-json`, but others as well) were failing before we ever discovered a compiler to attempt publishing with because we attempted to solve and compile them with the current index. To fix them we have to solve with both the legacy and current indices and try both; we bail out early if either solution contains packages not in the current index.

New results:

```
--------------------
PUBLISH FAILURES
--------------------

983 out of 1443 packages had at least 1 version fail (507 packages had all versions fail or 35%)
6359 out of 10695 versions failed or 59%

  - Dependency compiler conflict: 1
  - Publishing failed: 92
  - No compilers usable for publishing: 567
  - Solving failed (dependencies): 5699
```

As usual here are the relevant files:

* Full list of publish failures with reasons: [publish-failures.json](https://github.com/purescript/registry-dev/files/13626233/publish-failures.json)
* Packages that had no successful versions: [removed-packages.txt](https://github.com/purescript/registry-dev/files/13626234/removed-packages.txt)
* Metadata produced by this run, with compilers associated: [metadata.zip](https://github.com/purescript/registry-dev/files/13626235/metadata.zip)
*  Manifest index produced by this run, with dependencies all fixed: [registry-index.zip](https://github.com/purescript/registry-dev/files/13626266/registry-index.zip)

thomashoneyman commented:
We have decided to manually fix the manifests for deku, bolson, and rito so that these widely-used packages are not dropped. That's available in the latest commit.

thomashoneyman commented:
As discussed in the PureScript chat, the latest commit makes a small tweak to preserve packages that are from the core org or its derivatives, or which has had a tag since the 0.13 release date (May 29, 2019).

f-f commented:
I had a look at the above files, and they look good to me - the list of packages that we'll reserve is minimal and meaningful, and the list of ~450 freed packages seems sensible (I have checked quite a few manually and they all seem to be old enough to not be worth supporting, as people won't be able to build with them anyways)

I started looking at the code but it's a sizeable patch so it will take a few days

thomashoneyman commented:
Let me know if you encounter tricky bits and would like an explanation. Also happy to jump on a quick call and walk through sections of the code.

f-f commented:
Thanks! I think the logistics of that will be tricky over the Christmas days, but I let's see if that works for next week

f-f commented on app/src/App/API.purs:
```diff
@@ -1038,7 +1045,7 @@ publishToPursuit { packageSourceDir, dependenciesDir, compiler, resolutions } =
       Except.throw $ "Could not publish your package to Pursuit because an error was encountered (cc: @purescript/packaging): " <> error
     Right _ ->
-      Comment.comment "Successfully uploaded package docs to Pursuit! 🎉 🚀"
+      FS.Extra.remove tmp
```
We don't notify anymore for successful docs publishing?

thomashoneyman replied:
We do, the comment is just done outside of this function now. See e.g. line 581.

thomashoneyman commented on SPEC.md:
```diff
@@ -234,11 +234,12 @@ For example:
 - `hash`: a [`Sha256`](#Sha256) of the compressed archive fetched by the registry for the given version
 - `bytes`: the size of the tarball in bytes
 - `publishedTime`: the time the package was published as an `ISO8601` string
+- `compilers`: compiler versions this package is known to work with. This field can be in one of two states: a single version indicates that the package worked with a specific compiler on upload but has not yet been tested with all compilers, whereas a non-empty array of versions indicates the package has been tested with all compilers the registry supports.
```
The issue with the non-empty array is that it isn't clear whether an array of a single element represents one of:

* a package that has been published with the given compiler, but which hasn't been tested against the full set of compilers
* a package that has been tested against the full set of compilers and only works with one

f-f replied:
When are we going to end up in a situation where we don't test the package against the whole set of compilers? My reading of the PR is that we always do?

In any case, we'll always have packages that are not "tested against the full set of compilers": when a new compiler version comes out, then all packages will need a retest, and if a package doesn't have the new compiler in the array then we don't know if it's not compatible or if it hasn't been tested yet.

Maybe we need another piece of state somewhere else?

thomashoneyman replied:
> When are we going to end up in a situation where we don't test the package against the whole set of compilers? My reading of the PR is that we always do?

Yes, as implemented here we just go ahead and test everything as soon as we've published. However, I split out the state because in our initial discussions we worried about how long it takes for the compiler builds to run (it takes publishing from N seconds to N minutes in some cases — large libraries or ones that leverage a lot of type machinery). We'd originally talked about the compiler matrix being a cron job that runs later in the day. I just made it part of the publishing pipeline directly because it was simpler to implement.

If we decide that it's OK for publishing to take a long time then we can eliminate this state and just test the compilers immediately. In that case we'd just have a non-empty array.

thomashoneyman replied:
> In any case, we'll always have packages that are not "tested against the full set of compilers": when a new compiler version comes out, then all packages will need a retest, and if a package doesn't have the new compiler in the array then we don't know if it's not compatible or if it hasn't been tested yet.

Yea, that's a good point. You don't know if the metadata you're reading just hasn't been reached yet by an ongoing mass compiler build to check a new compiler.

> Maybe we need another piece of state somewhere else?

Off the top of my head I don't know a good place to put some state about possible compiler support; the metadata files are not helpful if a new compiler comes out and we're redoing the build since they're only aware of the one package.

thomashoneyman replied:
> If we decide that it's OK for publishing to take a long time then we can eliminate this state and just test the compilers immediately. In that case we'd just have a non-empty array.

I'm cool with this if you are.

> We'll always have packages that are not "tested against the full set of compilers" [...] maybe we need another piece of state somewhere else?

We could either a) say that the supported list of compilers for a package can potentially be missing the current compiler if the matrix is currently running and not bother with state or b) put a JSON file or something in the metadata directory that indicates whether the compiler matrix is running. Then consumers can look at that.

Personally the matrix runs infrequently enough (just new compiler releases!) that I would rather opt for (a).

f-f replied:
I pondered this for a few days and I think it's complicated?

Since we're going towards a model where we'd only run one registry job at a time and queue the rest (to prevent concurrent pushes to the repo), I'm afraid that running the whole matrix at once would make publishing very slow.
Something that we could do to counteract this could be to split the "publish" and the "matrix runs": on publishing we'd just add the package metadata with one compiler, and at the end of the publishing job we'd queue a series of "compiler matrix" jobs, each testing one compiler. These jobs would be of low priority, so new publishes would get in front of the queue, and things can stay snappy.

> Personally the matrix runs infrequently enough (just new compiler releases!) that I would rather opt for (a).

The approach detailed above implies that we're in a world where we do (a), i.e. the list of compilers is always potentially out of date, and that's fine.

f-f replied:
Additional note about the above: since the above would be introducing an "asynchronous matrix builder", we need to consider the dependency tree in our rebuilding: if a package A is published with compiler X, and then a package B depending on it is immediately published after it (a very common usecase since folks seem to publish their packages in batches), then we'd need to either make sure that matrix-build jobs for B are always run after matrix-build jobs for A, or retry them somehow.

thomashoneyman commented on SPEC.md:
```diff
@@ -234,11 +234,12 @@ For example:
 All packages in the registry have an associated metadata file, which is located in the `metadata` directory of the `registry` repository under the package name. For example, the metadata for the `aff` package is located at: https://github.com/purescript/registry/blob/main/metadata/aff.json. Metadata files are the source of truth on all published and unpublished versions for a particular package f ...
 
-Each published version of a package records three fields:
+Each published version of a package records four fields:
```
Yea, I think you're right 😆

thomashoneyman commented on lib/src/Metadata.purs:
```diff
@@ -69,18 +72,33 @@ codec = Profunctor.wrapIso Metadata $ CA.object "Metadata"
   , hash :: Sha256
   , publishedTime :: DateTime
+
+  -- UNSPECIFIED: Will be removed in the future.
```
I think it's because we only need it when we have the importer running off of Git tags but I don't fully remember either. I'm still in favor of recording the full location of a package in each published version so we can always reconstruct where it came from.

thomashoneyman replied:
https://github.com/purescript/registry-dev/issues/527 — it's for legacy package sets

thomashoneyman commented:
@f-f I think I've responded to your comments above, but GitHub isn't showing all of them in this main PR view so I'm not completely sure. Please let me know if I missed one!

thomashoneyman commented:
@f-f Did you have any other questions or comments about this work?

f-f commented:
@thomashoneyman I am through with the review - the overall shape of the code is good, and the only thing left to really resolve is the thing about "testing against the full set of compilers"

thomashoneyman commented:
Resolved merge conflicts, but I'm not sure what's happening with the integration test. Unfortunately I no longer have a NixOS system at home (though it's still Linux), and the integration test locally fails to git clone effect due to the tmp directory already containing it :shrug: I'll have to debug that later.

thomashoneyman commented:
Integration test was already passing in CI, but it was failing for me locally. Diagnosed the issue as arising from the `retryWithTimeout`, which would occasionally kill the clone if it was running slowly, leaving files behind, but those files weren't cleaned up well before the next retry. I've bumped the timeout and added a `cleanupOnCancel` function to the retry helper so we can do a cleanup prior to the next retry.

thomashoneyman commented:
Results of the most recent run:

```
1460 of 1790 total packages were considered for publishing (others had no manifests imported.)
  - 291 out of 1460 packages fully succeeded.
  - 994 packages partially succeeded.
  - 51 packages fully failed, but are reserved due to 0.13 or organization status.
  - 454 packages had all versions fail and will be removed.

10906 of 13282 total versions were considered for publishing.
  - 6363 out of 13282 versions failed.

    - Publishing failed: 85
    - No compilers usable for publishing: 562
    - Solving failed (dependencies): 5716
```

thomashoneyman commented:
I should emphasize too that if this merges we will need to do the reupload right away, because in compilation the solver will consider package dependencies with the compiler version included; if no packages have any compiler versions then new uploads are going to fail

f-f commented:
Right, thanks for the clarifications @thomashoneyman!

So the plan is:
- I am running the legacy importer locally to build a cache
- I will now have a look at the Spago codebase to see if we're going to need any updates to deal with this
- once the dry-run for the importer is done (in 24h or so) I will have a cache, so I'll reupload the registry and merge this branch. I'll need to think of what I need to disable to prevent uploads while I do the regen.

thomashoneyman commented:
You would need to disable the GitHub actions on the registry repository and sleep/kill the server while you reupload

thomashoneyman commented on app/src/App/SQLite.purs:
```diff
@@ -206,7 +207,7 @@ selectJob db { level: maybeLogLevel, since, jobId: JobId jobId } = do
     , selectMatrixJobById logs
     , selectTransferJobById logs
-    , selectPackageSetJobById logs
+    , selectAdminJobById logs
```
Is this our priority ordering? Seems like unpublish should be up at the top — or am I misinterpreting?

f-f replied:
This is just machinery to select the right job while aggregating from different tables (and part of the reason why I don't really want any more tables).

The priority ordering is coming from the order in which we pull jobs (admittedly in a noisy way) from the tables, here:
https://github.com/purescript/registry-dev/blob/b81f6522784f72addf31ad132812857b195b1ac3/app/src/App/Server/JobExecutor.purs#L102-L107

f-f commented on lib/src/API/V1.purs:
```diff
@@ -149,9 +156,34 @@ type MatrixJobData = JobInfo
+  = AdminPackageTransfer
+  | AdminLegacyImport LegacyImportMode
+  | AdminPackageSetUpdate PackageSetUpdateMode
+  | AdminPackageSetOperation PackageSetOperation -- For manual API requests
```
Agreed - I considered it briefly and probably would have reworked it eventually, but indeed we don't want this stuff to be public API.

However: they are here because they can be returned from the API, since they are in the database, and have a JobId, etc. Should we hide them from the world? It might be useful to have all this data when we go visualize all this stuff in a dashboard for our monitoring.

f-f commented:
Agreed - I think turning these special scripts into "routines that enqueue jobs" is the neatest option we have.

For the importer job: I would like us to have a daily publish job rather than say a GH action; it's not that much effort from us (we have them coded already), it helps smooth out the migration now and even when we stop grandfathering in legacy packages it still reduces the burden on maintainers, as they can publish once, and the subsequent package versions will be published without them having to do any manual intervention (if their repo is single-package, and if they use tags, and if their tags respect the naming convention. Overall a pretty strict set of conditions, and I expect we'll want to monitor how many packages are imported in this way over time)

About the legacy dependency parsing for the importer: I wonder if we can allow the pipeline to use those only when the publish job is signed by pacchettibotti (i.e. coming from the server itself), like we do for the package set jobs. I need to see how this shakes out in the implementation, I am sure there are some details missing. For example, we'd need to keep the signature around when calling the publish job because we need to verify it inside the job (vs just checking at the API layer for package set jobs)

In any case, I will revert my latest commit that introduced the admin jobs and explore this new approach in a separate PR to this branch

garyb commented:
There are some surprising things to me in `reserved-packages`, although I didn't notice anything suspicious in `removed-packages`.

`json` being an example, it's listed in `publish` failures with this:

```
  "json": {
    "1.0.0": {
      "reason": "Some resolutions from legacy index are not registered: arrays@7.3.0, bifunctors@6.1.0, const@6.0.0, contravariant@6.0.0, control@6.0.0, distributive@6.0.0, either@6.1.0, enums@6.0.1, exists@6.0.0, foldable-traversable@6.0.0, functors@5.0.0, gen@4.0.0, identity@6.0.0, integers@6.0.0, invariant@6.0.0, maybe@6.0.0, newtype@5.0.0, nonempty@7.0.0, numbers@9.0.1, orders@6.0.0, profunctor@6.0.1, safe-coerce@2.0.0, st@6.2.0, strings@6.0.1, tailrec@6.1.0, tuples@7.0.0, unfoldable@6.0.0, unsafe-coerce@6.0.0",
      "tag": "SolveFailedDependencies"
    },
    "1.1.0": {
      "reason": "Some resolutions from legacy index are not registered: arrays@7.3.0, bifunctors@6.1.0, const@6.0.0, contravariant@6.0.0, control@6.0.0, distributive@6.0.0, either@6.1.0, enums@6.0.1, exists@6.0.0, foldable-traversable@6.0.0, functors@5.0.0, gen@4.0.0, identity@6.0.0, integers@6.0.0, invariant@6.0.0, maybe@6.0.0, newtype@5.0.0, nonempty@7.0.0, numbers@9.0.1, orders@6.0.0, profunctor@6.0.1, safe-coerce@2.0.0, st@6.2.0, strings@6.0.1, tailrec@6.1.0, tuples@7.0.0, unfoldable@6.0.0, unsafe-coerce@6.0.0",
      "tag": "SolveFailedDependencies"
    }
  }
```
These are all core libraries, so not sure why... although actually that just led me into a new potential problem, I took a look at `publish-failures` for `arrays`, and it has a similar error for `7.3.0`... which is the latest published version of `arrays`, so maybe there's another report that needs generating there and considering - packages where the latest version is going to become unavailable also?

garyb commented:
Actually, following a chain of these core library version failures I eventually came to this:

```
  "unsafe-coerce": {
    "6.0.0": {
      "reason": "Failed to download unsafe-coerce@6.0.0 from the registry archive: HTTP status 404",
      "tag": "PublishError"
    }
  }
```

So perhaps a publish error that happened for that somehow caused this chain reaction of other failures now we're trying to rebuild the registry?
